### PR TITLE
[SPARK-8484] [ML]. Added TrainValidationSplit for hyper-parameter tuning. 

### PR DESCRIPTION
**Pull request by zapletal-martin**
- [X] Added TrainValidationSplit for hyper-parameter tuning. It randomly splits the input dataset into train and validation and use evaluation metric on the validation set to select the best model. It should be similar to CrossValidator, but simpler and less expensive.
- [X] Refactored CrossValidator to have both validators share code
- [X] External API of CrossValidator should stay unchanged

### CODE REVIEW

**Comment by mengxr** [comment]
@harsha2010 Do you have time to help review this PR? Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+import org.apache.spark.Logging
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.ml.evaluation.Evaluator
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.ml.{Model, Estimator}
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.types.StructType
+
+import scala.reflect.ClassTag
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala]: sort imports in the order specified in the style guide here:
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports

Comment (sujkh85) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala]: NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8484] [ML]. Added TrainValidationSplit for hyper-parameter tuning. (#6996)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

File: mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  override def fit(dataset: DataFrame): M = {
+    val sqlCtx = dataset.sqlContext
+    val est = $(estimator)
+    val eval = $(evaluator)
+    val epm = $(estimatorParamMaps)
+    val numModels = epm.length
+
+    val metrics = validationLogic(dataset, est, eval, epm, numModels)
+
+    logInfo(s"Average validation metrics: ${metrics.toSeq}")
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala]: why average? isn't this dumping all the validation metrics?

File: mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala [hunk 3]
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.mllib.util
 
+import org.apache.spark.util.Utils
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala]: organize imports

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Params for [[TrainValidationSplit]] and [[TrainValidationSplitModel]].
+ */
+private[ml] trait TrainValidationSplitParams extends ValidationParams {
+  /**
+   * Param for ratio between train and validation data. Must be between 0 and 1.
+   * Default: 0.75
+   * @group param
+   */
+  val trainRatio: DoubleParam = new DoubleParam(this, "numFolds",
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: why is the trainRatio called numFolds?

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 5]
```diff
@@ -91,74 +60,44 @@ class CrossValidator(override val uid: String) extends Estimator[CrossValidatorM
...
     val splits = MLUtils.kFold(dataset.rdd, $(numFolds), 0)
+
     splits.zipWithIndex.foreach { case ((training, validation), splitIndex) =>
       val trainingDataset = sqlCtx.createDataFrame(training, schema).cache()
       val validationDataset = sqlCtx.createDataFrame(validation, schema).cache()
-      // multi-model training
       logDebug(s"Train split $splitIndex with multiple sets of parameters.")
-      val models = est.fit(trainingDataset, epm).asInstanceOf[Seq[Model[_]]]
-      trainingDataset.unpersist()
+      val newMetrics = measureModels(trainingDataset, validationDataset, est, eval, epm, numModels)
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala]: perhaps a more descriptive name than newMetrics might be metricsPerSplit? or something better.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    bestModel: Model[_],
+    avgMetrics: Array[Double])
+  extends ValidationModel[TrainValidationSplitModel](uid, bestModel, avgMetrics)
+  with TrainValidationSplitParams {
+
+  override def copy(extra: ParamMap): TrainValidationSplitModel = {
+    val copied = new TrainValidationSplitModel (
+      uid,
+      bestModel.copy(extra).asInstanceOf[Model[_]],
+      avgMetrics.clone())
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: i'm not sure if array.clone() is considered standard practice in Scala.
arrays are handled specially so even though the clone method throws an error it is just a stub, but it is still a 
bit confusing i think to use array.clone() instead of say Array.copy(...)

File: mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.types.StructType
+
+
+/**
+ * :: DeveloperApi ::
+ * Common params for [[TrainValidationSplitParams]] and [[CrossValidatorParams]].
+ */
+@DeveloperApi
+private[ml] trait ValidationParams extends Params {
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/Validation.scala]: Can we name this ValidatorParams and Validator to be consistent?

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+      est: Estimator[_],
+      eval: Evaluator,
+      epm: Array[ParamMap],
+      numModels: Int): Array[Double] = {
+
+    val schema = dataset.schema
+    transformSchema(schema, logging = true)
+    val sqlCtx = dataset.sqlContext
+
+    val splits = MLUtils.sample(dataset.rdd, $(trainRatio), 1)
```
Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: can use tuple destructuring here instead of ._1, ._2
val (training, validation) = MLUtils.sample(...)

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala (hunk 6 above)
Comment (zapletal-martin) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: I actually took this code from CrossValidator without thinking about it. But I think clone() should be fine. From what I read there isn't a big difference. copy() is a bit faster, but since the array will only be small then I assume clone() should be fine and we can avoid a line or two of boilerplate.

Comment (harsha2010) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala]: @zapletal-martin  make sense..i was not worried about performance as much as readability. if clone was already being used then it should be fine

File: mllib/src/main/scala/org/apache/spark/ml/tuning/Validator.scala [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    val eval = $(evaluator)
+    val epm = $(estimatorParamMaps)
+    val numModels = epm.length
+
+    val metrics = validationLogic(dataset, est, eval, epm, numModels)
+
+    logInfo(s"Average validation metrics: ${metrics.toSeq}")
+    val (bestMetric, bestIndex) = metrics.zipWithIndex.maxBy(_._1)
+    logInfo(s"Best set of parameters:\n${epm(bestIndex)}")
+    logInfo(s"Best cross-validation metric: $bestMetric.")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/Validator.scala]: Same here. `cross-validation` would appear under `TrainValidationSplit`.

**Comment by mengxr** [comment]
@zapletal-martin Sorry for my late comment! But this PR contains much more content than I expected. We should try to keep each PR minimal. For example, implementing `TrainValidationSplit` alone would make the PR much easier to review. Using `ValidatorParams` for both `CrossValidator` and `TrainValidationSplit` sounds good to me. But the `Validator` refactoring could be done in a follow-up PR, and the changes to `MLUtils` could be in a separate PR as well. Does it sound good to you?

**Comment by zapletal-martin** [comment]
@mengxr sure, sounds good. There was quite a lot of code duplication so I decided to refactor it. I will create a new PR just for TrainValidationSplit without the refactor and we can address the resulting duplication in a later PR.

**Comment by mengxr** [comment]
@zapletal-martin We can use `ValidatorParams` in the new PR, just without `Validator` refactoring and `MLUtils` changes. Thanks for your understanding!
